README.md
# yyl-ssr-logger

用于 yyl-ssr 项目 服务端 日志收集

## 安装

```bash
# yarn
yarn add yyl-ssr-logger

# npm
npm i yyl-ssr-logger --save
```

## 使用

```typescript
import path from 'path'
import { Log, LogType } from 'yyl-ssr-logger'
const iLog = new Log({
  logPath: path.join(__dirname, 'log')
})
iLog.log({ type: LogType.Info, path: '/path/to/request', args: ['hello worl'] })
```

## types

直接看 types 吧

```typescript
/** 日志类型 */
export declare enum LogType {
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
  Success = 'success',
  System = 'system'
}
/** 日志初始化配置 */
export interface LogOption<T> {
  /** 日志保存路径 */
  logPath?: string
  /** 运行日志存储上限 */
  runtimeLimitSize?: number
  /** 日志写入间隔 */
  writeInterval?: number
  /** 运行日志名称 */
  runtimeFilename?: string
  /** 错误日志名称 */
  errorFilename?: string
  /** 打印日志 */
  verbose?: boolean
  /** 格式化函数 */
  formatter?: (log: T & LogFormatOption) => string
  /** 日志分隔符 */
  logSep?: string
  /** debug 日志输出接受函数 */
  logger?: (type: LogType, args: any[]) => void
}
/** log 属性用 types */
export declare type LogProperty<T> = Required<LogOption<T>>
/** 日志入参 */
export interface LogArgu {
  /** 日志类型 */
  type?: LogType
  /** 请求路径 */
  path?: string
  /** 日志内容 */
  args: any[]
}
/** 日志 formatter 额外的参数 */
export interface LogFormatOption extends Required<LogArgu> {
  date: string
}
/** 日志对象 */
export declare class Log<T = {}> {
  /** 日志缓存 */
  private logCache
  /** interval key */
  private intervalKey
  /** 运行日志目录 */
  private runtimeLogPath
  /** 错误日志目录 */
  private errorLogPath
  /** 日志存储路径 */
  private logPath
  /** log 大小上限 */
  private runtimeLimitSize
  /** 日志写入间隔 */
  private writeInterval
  /** 打印日志 */
  private verbose
  /** 运行日志文件名 */
  private runtimeFilename
  /** 错误日志文件名 */
  private errorFilename
  /** 日志分隔符 */
  private logSep
  /** 日志解析器 */
  private formatter
  /** debug 日志接收器 */
  private logger
  constructor(option?: LogOption<T>)
  /** 日志记录 */
  log(op: T & LogArgu): void
  /** 日志写入操作 */
  private writer
  /** 日志清除 */
  clear(): void
}
```


history.md
# 历史版本

## 0.1.1 (2020-11-18)

- feat: 修改 content 输出到日志格式改为 string

## 0.1.0 (2020-11-16)

- feat: 诞生


test/jest/index.test.js
const { Log, LogType } = require('../../')
const path = require('path')
const extFs = require('yyl-fs')
const fs = require('fs')
const { waitFor } = require('yyl-util')

const FRAG_PATH = path.join(__dirname, '__frag')

test('usage test', async () => {
  const logPath = path.join(FRAG_PATH, 'index')
  const runtimePath = path.join(logPath, 'runtime.log')
  const errorPath = path.join(logPath, 'error.log')
  const logs = []
  await extFs.removeFiles(logPath)
  const iLog = new Log({
    logPath,
    verbose: true,
    logger(type, args) {
      logs.push([type, args])
    }
  })

  // + input
  iLog.log({
    type: LogType.Info,
    path: '/index',
    args: ['123', { zz: 'abc' }]
  })

  iLog.log({
    type: LogType.Error,
    path: '/index',
    args: ['123', new Error('456')]
  })
  // - input

  await waitFor(1000)
  // + check
  expect(fs.existsSync(runtimePath)).toEqual(true)
  expect(fs.existsSync(errorPath)).toEqual(true)
  expect(fs.readFileSync(runtimePath).toString().split('\n').length).toEqual(3)
  expect(fs.readFileSync(errorPath).toString().split('\n').length).toEqual(2)
  expect(logs.length).toEqual(3)
  // - check
})


output/index.js
/*!
 * yyl-ssr-logger cjs 0.1.1
 * (c) 2020 - 2020 jackness
 * Released under the MIT License.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var path = _interopDefault(require('path'));
var fs = _interopDefault(require('fs'));
var dayjs = _interopDefault(require('dayjs'));
var chalk = _interopDefault(require('chalk'));

function mkdirSync(toFile) {
    const tPath = toFile.replace(/[/\\]$/, '');
    const r = [];
    (function deep(iPath) {
        if (fs.existsSync(iPath) || /[/\\]$/.test(iPath)) ;
        else {
            deep(path.dirname(iPath));
            fs.mkdirSync(iPath);
            r.push(iPath);
        }
    })(tPath);
    return r;
}

(function (LogType) {
    LogType["Info"] = "info";
    LogType["Warn"] = "warn";
    LogType["Error"] = "error";
    LogType["Success"] = "success";
    LogType["System"] = "system";
})(exports.LogType || (exports.LogType = {}));
/** 日志对象 */
class Log {
    constructor(option) {
        /** 日志缓存 */
        this.logCache = [];
        /** interval key */
        this.intervalKey = 0;
        /** 运行日志目录 */
        this.runtimeLogPath = '';
        /** 错误日志目录 */
        this.errorLogPath = '';
        /** 日志存储路径 */
        this.logPath = path.join(process.cwd(), './log');
        /** log 大小上限 */
        this.runtimeLimitSize = 1000 * 1024 * 2;
        /** 日志写入间隔 */
        this.writeInterval = 1000;
        /** 打印日志 */
        this.verbose = false;
        /** 运行日志文件名 */
        this.runtimeFilename = 'runtime.log';
        /** 错误日志文件名 */
        this.errorFilename = 'error.log';
        /** 日志分隔符 */
        this.logSep = '\n';
        /** 日志解析器 */
        this.formatter = (log) => {
            const rLog = Object.assign({}, log);
            // 去掉 args 参数
            delete rLog.args;
            return JSON.stringify(rLog);
        };
        /** debug 日志接收器 */
        this.logger = (type, args) => {
            console.log(`${chalk.green('[ssr]')} - ${chalk[type === exports.LogType.Error ? 'red' : 'gray'](`[${type}]`)}`, ...args);
        };
        // 属性初始化
        if (option === null || option === void 0 ? void 0 : option.logPath) {
            this.logPath = option.logPath;
        }
        if ((option === null || option === void 0 ? void 0 : option.runtimeLimitSize) !== undefined) {
            this.runtimeLimitSize = option.runtimeLimitSize;
        }
        if (option === null || option === void 0 ? void 0 : option.writeInterval) {
            this.writeInterval = option.writeInterval;
        }
        if (option === null || option === void 0 ? void 0 : option.verbose) {
            this.verbose = option.verbose;
        }
        if (option === null || option === void 0 ? void 0 : option.runtimeFilename) {
            this.runtimeFilename = option.runtimeFilename;
        }
        if (option === null || option === void 0 ? void 0 : option.errorFilename) {
            this.errorFilename = option.errorFilename;
        }
        if (option === null || option === void 0 ? void 0 : option.formatter) {
            this.formatter = option.formatter;
        }
        if (option === null || option === void 0 ? void 0 : option.logger) {
            this.logger = option.logger;
        }
        // 运行日志目录初始化
        const runtimeLogPath = path.resolve(this.logPath, this.runtimeFilename);
        mkdirSync(path.dirname(runtimeLogPath));
        fs.writeFileSync(runtimeLogPath, '');
        this.runtimeLogPath = runtimeLogPath;
        // 错误日志目录初始化
        const errorLogPath = path.resolve(this.logPath, this.errorFilename);
        mkdirSync(path.dirname(errorLogPath));
        fs.writeFileSync(errorLogPath, '');
        this.errorLogPath = errorLogPath;
        // 启动定时任务
        this.intervalKey = setInterval(() => {
            this.writer();
        }, this.writeInterval);
    }
    /** 日志记录 */
    log(op) {
        var _a;
        const { logCache, verbose, logger } = this;
        const param = Object.assign(Object.assign({}, op), { type: op.type || exports.LogType.Info, path: op.path || 'system', content: ((_a = op.args) === null || _a === void 0 ? void 0 : _a.map((ctx) => {
                if (typeof ctx === 'object') {
                    if (typeof ctx.stack === 'string') {
                        return ctx.stack;
                    }
                    else {
                        return JSON.stringify(ctx);
                    }
                }
                else {
                    return ctx;
                }
            }).join(' ')) || '', date: dayjs().format('YYYY-MM-DD hh:mm:ss') });
        logCache.push(param);
        if (verbose) {
            logger(param.type, param.args);
        }
    }
    /** 日志写入操作 */
    writer() {
        const { logCache, runtimeLimitSize, runtimeLogPath, errorLogPath, formatter, logSep, verbose, logger } = this;
        if (!logCache.length) {
            return;
        }
        // 运行日志
        const runtimeLogs = logCache.map((log) => formatter(log));
        // 错误日志
        const errorLogs = logCache
            .filter((log) => log.type === exports.LogType.Error)
            .map((log) => formatter(log));
        // 运行日志写入
        if (runtimeLogs.length) {
            if (runtimeLimitSize > 0) {
                const runtimeSize = fs.statSync(runtimeLogPath).size;
                if (runtimeSize > runtimeLimitSize) {
                    // 超出减一半
                    const oriLogs = fs.readFileSync(runtimeLogPath).toString().split(logSep);
                    fs.writeFileSync(runtimeLogPath, `${oriLogs.join(logSep)}${logSep}${runtimeLogs.join(logSep)}${logSep}`);
                }
                else {
                    fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`);
                }
            }
            else {
                fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`);
            }
        }
        // 错误日志写入
        if (errorLogs.length) {
            fs.appendFileSync(errorLogPath, `${errorLogs.join(logSep)}${logSep}`);
        }
        if (verbose) {
            logger(exports.LogType.System, [
                '写入日志文件完成',
                `runtime: ${chalk.green(runtimeLogs.length)}`,
                `error: ${chalk.red(errorLogs.length)}`
            ]);
        }
        logCache.length = 0;
    }
    /** 日志清除 */
    clear() {
        this.logCache = [];
        clearInterval(this.intervalKey);
    }
}

exports.Log = Log;


rollup.config.js
import pkg from './package.json'
import typescript from 'rollup-plugin-typescript2'
import commonjs from 'rollup-plugin-commonjs'
import nodeResolve from 'rollup-plugin-node-resolve'
import json from 'rollup-plugin-json'

function buildBanner(type) {
  return [
    '/*!',
    ` * ${pkg.name} ${type} ${pkg.version}`,
    ` * (c) 2020 - ${new Date().getFullYear()} jackness`,
    ' * Released under the MIT License.',
    ' */'
  ].join('\n')
}

const IS_PUBLISH = process.env.NODE_ENV === 'production'

const config = {
  input: './src/index.ts',
  output: [],
  plugins: [
    nodeResolve({ jsnext: true }),
    commonjs(),
    json(),
    typescript({
      typescript: require('typescript')
    })
  ],
  external: ['dayjs', 'chalk', 'path', 'fs']
}

export default [
  {
    input: config.input,
    output: [
      {
        file: './output/index.js',
        format: 'cjs',
        banner: buildBanner('cjs'),
        exports: 'named',
        sourcemap: false
      }
    ],
    plugins: config.plugins,
    external: config.external
  },
  {
    input: config.input,
    output: [
      {
        file: './output/index.esm.js',
        format: 'esm',
        banner: buildBanner('esm'),
        sourcemap: false
      }
    ],
    plugins: config.plugins,
    external: config.external
  }
]


output/index.d.ts
/** 日志类型 */
export declare enum LogType {
    Info = "info",
    Warn = "warn",
    Error = "error",
    Success = "success",
    System = "system"
}
/** 日志初始化配置 */
export interface LogOption<T> {
    /** 日志保存路径 */
    logPath?: string;
    /** 运行日志存储上限 */
    runtimeLimitSize?: number;
    /** 日志写入间隔 */
    writeInterval?: number;
    /** 运行日志名称 */
    runtimeFilename?: string;
    /** 错误日志名称 */
    errorFilename?: string;
    /** 打印日志 */
    verbose?: boolean;
    /** 格式化函数 */
    formatter?: (log: T & LogFormatOption) => string;
    /** 日志分隔符 */
    logSep?: string;
    /** debug 日志输出接受函数 */
    logger?: (type: LogType, args: any[]) => void;
}
/** log 属性用 types */
export declare type LogProperty<T> = Required<LogOption<T>>;
/** 日志入参 */
export interface LogArgu {
    /** 日志类型 */
    type?: LogType;
    /** 请求路径 */
    path?: string;
    /** 日志内容 */
    args: any[];
}
/** 日志 formatter 额外的参数 */
export interface LogFormatOption extends Required<LogArgu> {
    date: string;
}
/** 日志对象 */
export declare class Log<T = {}> {
    /** 日志缓存 */
    private logCache;
    /** interval key */
    private intervalKey;
    /** 运行日志目录 */
    private runtimeLogPath;
    /** 错误日志目录 */
    private errorLogPath;
    /** 日志存储路径 */
    private logPath;
    /** log 大小上限 */
    private runtimeLimitSize;
    /** 日志写入间隔 */
    private writeInterval;
    /** 打印日志 */
    private verbose;
    /** 运行日志文件名 */
    private runtimeFilename;
    /** 错误日志文件名 */
    private errorFilename;
    /** 日志分隔符 */
    private logSep;
    /** 日志解析器 */
    private formatter;
    /** debug 日志接收器 */
    private logger;
    constructor(option?: LogOption<T>);
    /** 日志记录 */
    log(op: T & LogArgu): void;
    /** 日志写入操作 */
    private writer;
    /** 日志清除 */
    clear(): void;
}


output/index.esm.js
/*!
 * yyl-ssr-logger esm 0.1.1
 * (c) 2020 - 2020 jackness
 * Released under the MIT License.
 */
import path from 'path';
import fs from 'fs';
import dayjs from 'dayjs';
import chalk from 'chalk';

function mkdirSync(toFile) {
    const tPath = toFile.replace(/[/\\]$/, '');
    const r = [];
    (function deep(iPath) {
        if (fs.existsSync(iPath) || /[/\\]$/.test(iPath)) ;
        else {
            deep(path.dirname(iPath));
            fs.mkdirSync(iPath);
            r.push(iPath);
        }
    })(tPath);
    return r;
}

/** 日志类型 */
var LogType;
(function (LogType) {
    LogType["Info"] = "info";
    LogType["Warn"] = "warn";
    LogType["Error"] = "error";
    LogType["Success"] = "success";
    LogType["System"] = "system";
})(LogType || (LogType = {}));
/** 日志对象 */
class Log {
    constructor(option) {
        /** 日志缓存 */
        this.logCache = [];
        /** interval key */
        this.intervalKey = 0;
        /** 运行日志目录 */
        this.runtimeLogPath = '';
        /** 错误日志目录 */
        this.errorLogPath = '';
        /** 日志存储路径 */
        this.logPath = path.join(process.cwd(), './log');
        /** log 大小上限 */
        this.runtimeLimitSize = 1000 * 1024 * 2;
        /** 日志写入间隔 */
        this.writeInterval = 1000;
        /** 打印日志 */
        this.verbose = false;
        /** 运行日志文件名 */
        this.runtimeFilename = 'runtime.log';
        /** 错误日志文件名 */
        this.errorFilename = 'error.log';
        /** 日志分隔符 */
        this.logSep = '\n';
        /** 日志解析器 */
        this.formatter = (log) => {
            const rLog = Object.assign({}, log);
            // 去掉 args 参数
            delete rLog.args;
            return JSON.stringify(rLog);
        };
        /** debug 日志接收器 */
        this.logger = (type, args) => {
            console.log(`${chalk.green('[ssr]')} - ${chalk[type === LogType.Error ? 'red' : 'gray'](`[${type}]`)}`, ...args);
        };
        // 属性初始化
        if (option === null || option === void 0 ? void 0 : option.logPath) {
            this.logPath = option.logPath;
        }
        if ((option === null || option === void 0 ? void 0 : option.runtimeLimitSize) !== undefined) {
            this.runtimeLimitSize = option.runtimeLimitSize;
        }
        if (option === null || option === void 0 ? void 0 : option.writeInterval) {
            this.writeInterval = option.writeInterval;
        }
        if (option === null || option === void 0 ? void 0 : option.verbose) {
            this.verbose = option.verbose;
        }
        if (option === null || option === void 0 ? void 0 : option.runtimeFilename) {
            this.runtimeFilename = option.runtimeFilename;
        }
        if (option === null || option === void 0 ? void 0 : option.errorFilename) {
            this.errorFilename = option.errorFilename;
        }
        if (option === null || option === void 0 ? void 0 : option.formatter) {
            this.formatter = option.formatter;
        }
        if (option === null || option === void 0 ? void 0 : option.logger) {
            this.logger = option.logger;
        }
        // 运行日志目录初始化
        const runtimeLogPath = path.resolve(this.logPath, this.runtimeFilename);
        mkdirSync(path.dirname(runtimeLogPath));
        fs.writeFileSync(runtimeLogPath, '');
        this.runtimeLogPath = runtimeLogPath;
        // 错误日志目录初始化
        const errorLogPath = path.resolve(this.logPath, this.errorFilename);
        mkdirSync(path.dirname(errorLogPath));
        fs.writeFileSync(errorLogPath, '');
        this.errorLogPath = errorLogPath;
        // 启动定时任务
        this.intervalKey = setInterval(() => {
            this.writer();
        }, this.writeInterval);
    }
    /** 日志记录 */
    log(op) {
        var _a;
        const { logCache, verbose, logger } = this;
        const param = Object.assign(Object.assign({}, op), { type: op.type || LogType.Info, path: op.path || 'system', content: ((_a = op.args) === null || _a === void 0 ? void 0 : _a.map((ctx) => {
                if (typeof ctx === 'object') {
                    if (typeof ctx.stack === 'string') {
                        return ctx.stack;
                    }
                    else {
                        return JSON.stringify(ctx);
                    }
                }
                else {
                    return ctx;
                }
            }).join(' ')) || '', date: dayjs().format('YYYY-MM-DD hh:mm:ss') });
        logCache.push(param);
        if (verbose) {
            logger(param.type, param.args);
        }
    }
    /** 日志写入操作 */
    writer() {
        const { logCache, runtimeLimitSize, runtimeLogPath, errorLogPath, formatter, logSep, verbose, logger } = this;
        if (!logCache.length) {
            return;
        }
        // 运行日志
        const runtimeLogs = logCache.map((log) => formatter(log));
        // 错误日志
        const errorLogs = logCache
            .filter((log) => log.type === LogType.Error)
            .map((log) => formatter(log));
        // 运行日志写入
        if (runtimeLogs.length) {
            if (runtimeLimitSize > 0) {
                const runtimeSize = fs.statSync(runtimeLogPath).size;
                if (runtimeSize > runtimeLimitSize) {
                    // 超出减一半
                    const oriLogs = fs.readFileSync(runtimeLogPath).toString().split(logSep);
                    fs.writeFileSync(runtimeLogPath, `${oriLogs.join(logSep)}${logSep}${runtimeLogs.join(logSep)}${logSep}`);
                }
                else {
                    fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`);
                }
            }
            else {
                fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`);
            }
        }
        // 错误日志写入
        if (errorLogs.length) {
            fs.appendFileSync(errorLogPath, `${errorLogs.join(logSep)}${logSep}`);
        }
        if (verbose) {
            logger(LogType.System, [
                '写入日志文件完成',
                `runtime: ${chalk.green(runtimeLogs.length)}`,
                `error: ${chalk.red(errorLogs.length)}`
            ]);
        }
        logCache.length = 0;
    }
    /** 日志清除 */
    clear() {
        this.logCache = [];
        clearInterval(this.intervalKey);
    }
}

export { Log, LogType };


src/index.ts
import path from 'path'
import fs from 'fs'
import { mkdirSync } from './lib/mkdirSync'
import dayjs from 'dayjs'
import chalk from 'chalk'

/** 日志类型 */
export enum LogType {
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
  Success = 'success',
  System = 'system'
}

/** 日志初始化配置 */
export interface LogOption<T> {
  /** 日志保存路径 */
  logPath?: string
  /** 运行日志存储上限 */
  runtimeLimitSize?: number
  /** 日志写入间隔 */
  writeInterval?: number
  /** 运行日志名称 */
  runtimeFilename?: string
  /** 错误日志名称 */
  errorFilename?: string
  /** 打印日志 */
  verbose?: boolean
  /** 格式化函数 */
  formatter?: (log: T & LogFormatOption) => string
  /** 日志分隔符 */
  logSep?: string
  /** debug 日志输出接受函数 */
  logger?: (type: LogType, args: any[]) => void
}

/** log 属性用 types */
export type LogProperty<T> = Required<LogOption<T>>

/** 日志入参 */
export interface LogArgu {
  /** 日志类型 */
  type?: LogType
  /** 请求路径 */
  path?: string
  /** 日志内容 */
  args: any[]
}

/** 日志 formatter 额外的参数 */
export interface LogFormatOption extends Required<LogArgu> {
  date: string
}

/** 日志对象 */
export class Log<T = {}> {
  /** 日志缓存 */
  private logCache: (T & Required<LogFormatOption>)[] = []
  /** interval key */
  private intervalKey: any = 0
  /** 运行日志目录 */
  private runtimeLogPath: string = ''
  /** 错误日志目录 */
  private errorLogPath: string = ''
  /** 日志存储路径 */
  private logPath: LogProperty<T>['logPath'] = path.join(process.cwd(), './log')
  /** log 大小上限 */
  private runtimeLimitSize: LogProperty<T>['runtimeLimitSize'] = 1000 * 1024 * 2
  /** 日志写入间隔 */
  private writeInterval: LogProperty<T>['writeInterval'] = 1000
  /** 打印日志 */
  private verbose: LogProperty<T>['verbose'] = false
  /** 运行日志文件名 */
  private runtimeFilename: LogProperty<T>['runtimeFilename'] = 'runtime.log'
  /** 错误日志文件名 */
  private errorFilename: LogProperty<T>['errorFilename'] = 'error.log'
  /** 日志分隔符 */
  private logSep: LogProperty<T>['logSep'] = '\n'
  /** 日志解析器 */
  private formatter: LogProperty<T>['formatter'] = (log) => {
    const rLog: Partial<T & LogFormatOption> = Object.assign({}, log)
    // 去掉 args 参数
    delete rLog.args
    return JSON.stringify(rLog)
  }

  /** debug 日志接收器 */
  private logger: LogProperty<T>['logger'] = (type, args) => {
    console.log(
      `${chalk.green('[ssr]')} - ${chalk[type === LogType.Error ? 'red' : 'gray'](`[${type}]`)}`,
      ...args
    )
  }

  constructor(option?: LogOption<T>) {
    // 属性初始化
    if (option?.logPath) {
      this.logPath = option.logPath
    }
    if (option?.runtimeLimitSize !== undefined) {
      this.runtimeLimitSize = option.runtimeLimitSize
    }
    if (option?.writeInterval) {
      this.writeInterval = option.writeInterval
    }
    if (option?.verbose) {
      this.verbose = option.verbose
    }
    if (option?.runtimeFilename) {
      this.runtimeFilename = option.runtimeFilename
    }
    if (option?.errorFilename) {
      this.errorFilename = option.errorFilename
    }
    if (option?.formatter) {
      this.formatter = option.formatter
    }

    if (option?.logger) {
      this.logger = option.logger
    }

    // 运行日志目录初始化
    const runtimeLogPath = path.resolve(this.logPath, this.runtimeFilename)
    mkdirSync(path.dirname(runtimeLogPath))
    fs.writeFileSync(runtimeLogPath, '')
    this.runtimeLogPath = runtimeLogPath

    // 错误日志目录初始化
    const errorLogPath = path.resolve(this.logPath, this.errorFilename)
    mkdirSync(path.dirname(errorLogPath))
    fs.writeFileSync(errorLogPath, '')
    this.errorLogPath = errorLogPath

    // 启动定时任务
    this.intervalKey = setInterval(() => {
      this.writer()
    }, this.writeInterval)
  }

  /** 日志记录 */
  log(op: T & LogArgu) {
    const { logCache, verbose, logger } = this
    const param = {
      ...op,
      type: op.type || LogType.Info,
      path: op.path || 'system',
      content:
        op.args
          ?.map((ctx) => {
            if (typeof ctx === 'object') {
              if (typeof ctx.stack === 'string') {
                return ctx.stack
              } else {
                return JSON.stringify(ctx)
              }
            } else {
              return ctx
            }
          })
          .join(' ') || '',
      date: dayjs().format('YYYY-MM-DD hh:mm:ss')
    }
    logCache.push(param)
    if (verbose) {
      logger(param.type, param.args)
    }
  }

  /** 日志写入操作 */
  private writer() {
    const {
      logCache,
      runtimeLimitSize,
      runtimeLogPath,
      errorLogPath,
      formatter,
      logSep,
      verbose,
      logger
    } = this
    if (!logCache.length) {
      return
    }
    // 运行日志
    const runtimeLogs = logCache.map((log) => formatter(log))
    // 错误日志
    const errorLogs = logCache
      .filter((log) => log.type === LogType.Error)
      .map((log) => formatter(log))

    // 运行日志写入
    if (runtimeLogs.length) {
      if (runtimeLimitSize > 0) {
        const runtimeSize = fs.statSync(runtimeLogPath).size
        if (runtimeSize > runtimeLimitSize) {
          // 超出减一半
          const oriLogs = fs.readFileSync(runtimeLogPath).toString().split(logSep)
          fs.writeFileSync(
            runtimeLogPath,
            `${oriLogs.join(logSep)}${logSep}${runtimeLogs.join(logSep)}${logSep}`
          )
        } else {
          fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`)
        }
      } else {
        fs.appendFileSync(runtimeLogPath, `${runtimeLogs.join(logSep)}${logSep}`)
      }
    }

    // 错误日志写入
    if (errorLogs.length) {
      fs.appendFileSync(errorLogPath, `${errorLogs.join(logSep)}${logSep}`)
    }

    if (verbose) {
      logger(LogType.System, [
        '写入日志文件完成',
        `runtime: ${chalk.green(runtimeLogs.length)}`,
        `error: ${chalk.red(errorLogs.length)}`
      ])
    }

    logCache.length = 0
  }

  /** 日志清除 */
  clear() {
    this.logCache = []
    clearInterval(this.intervalKey)
  }
}
